alternator/streams.cc

```diff
@@ -742,24 +829,51 @@ future<executor::request_return_type> executor::get_records(client_state& client
     auto& db = _proxy.get_db().local();
     schema_ptr schema, base;
     try {
-        auto& log_table = db.find_column_family(iter.shard.table);
+        auto& log_table = db.find_column_family(iter.table);
         schema = log_table.schema();
         base = cdc::get_base_table(db, *schema);
     } catch (...) {        
     }
 
     if (!schema || !base || !is_alternator_keyspace(schema->ks_name())) {
-        throw api_error::resource_not_found(boost::lexical_cast<std::string>(iter.shard.table));
+        throw api_error::resource_not_found(boost::lexical_cast<std::string>(iter.table));
+    }
+
+    auto f = make_ready_future<shard_iterator>(iter);
+    if (!iter.shard.has_expired()) {
+        // ugh. figure out if we have end-of-shard
+        // We try to avoid the extra query by using the
+        // gossip state as a cache check - i.e. know if
+        // topology has changed and invalidated the current
+        // iterator. This is far from exact, so we could still
+        // generate duplicate records at generation changes
```

Comment:
Is this a bug? a FIXME? or acceptable?

Reply:
I think it is an acceptable trade-off between correct and efficient. As I describe in reply to  your earlier comment, we need to check that the current CDC generation wasn't replaced. Doing a query into cdc tables is to expensive (me thinks), so we use this var as a check. Neither check (expensive or cheaper) would be 100% correct, since we may not see a new cdc gen until some time after it appears.

Reply:
I'm sorry, I guess I'm dense, I still don't understand... If this is "not 100% correct", isn't it... A bug? Which needs to be fixed? (and I didn't yet understand what this bug means).
Is this a bug caused by a CDC bug, or a new bug we're introducing especially for Alternator user?

Reply:
Correct as in "temporarily exact", i.e. just because we pass this check does not 100% mean we are not being replaced, just that we don't see it yet. I'll update the comment.